### PR TITLE
Add support for recvmmsg and UDP GRO

### PR DESCRIPTION
**Pull request by brbzull0**
### Descrition

This PR adds support for UDP GRO when reading the socket using `recvm(m)sg`,  also includes support for `recvmmsg` when it is available by the OS.

There is now a new configuration variable:

```yaml
# records.yaml
ts:
  udp:
    enable_gro: 1 # Enabled by default.
```
----

### Notes

All the tests were done using GET only and setting the h2load parameter to `--max-udp-payload-size=2k`.

|Feature       |Total # Req | # Clients | # recvmmsg | # recvmsg |
|--------------|------------|-----------|------------|-----------|
| recvmmsg+gro | 100k       | 100       |      22834 | 0         |
| recvmmsg     | 100k       | 100       |      23501 | 0         |
| recvmsg+gro  | 100k       | 100       |       | 247970         |
| recvmmsg     | 100k       | 100       |       | 249097         |

As I didn't see much of a difference between having GRO enabled or disabled I did run a test with debug
enabled and I was able to count how many packets came with the GRO set in the ancillary data. 
Then I realized why I wasn't actually seeing much difference, not too many packets were spliced by the kernel.

Packets with (No GRO/GRO) = Number of times that ATS detected a packet with spliced with GRO.

|Feature       |Total # Req | # Clients | # recvmmsg | # recvmsg | # Packets with (No GRO/GRO)  | GRO % |
|--------------|------------|-----------|------------|-----------|---------------|-------|
| recvmmsg+gro | 100k       | 100       |     13929  | 0         |    223080 / **180** | 0.8% |
| recvmsg+gro  | 100k       | 100       |      0| 278718         |    231760 / **34484** | 14% |

**Using recvmmsg reduced the number of syscall significantly**

This is what I used for this tests:

**bpftrace**

```bash
bpftrace -e 'tracepoint:syscalls:sys_enter_recvm* { @[comm, probe] = count();  }' 
```
**h2load**
```bash
/opt/bin/h2load -n 100000 -c 100 --npn-list=h3 URL --max-udp-payload-size=2k
```

I was using only **GET** requests with around 15 header with about 80 bytes each.
I have not tried POST yet which I believe will give us better insights.

I tested this on `Ubuntu 20.04` and `FreeBSD 13.1`. All figures here are from a tests on Ubuntu. 
This was tested on `FreeBSD 13.1` which does not support GSO/GRO to make sure the functionallity works the same.

### CODE REVIEW

**Comment by JosiahWI** [comment]
The CMake build will need to be updated as well.

**Comment by brbzull0** [comment]
> The CMake build will need to be updated as well.

Thanks for the comment. 
You mean this: https://github.com/apache/trafficserver/pull/9905/files#diff-49473dca262eeab3b4a43002adb08b4db31020d190caaad1594b47f1d5daa810R1393 ?

Cheers.

**Comment by JosiahWI** [comment]
I took a closer look; it looks like we don't check this yet in CMake, but we probably should. We might want to make our own function:
```cmake
include(CheckIncludeFile)
function(REQUIRE_SYSTEM_HEADER HEADER)
    check_include_file("${HEADER}" HAVE_HEADER)
    if(NOT HAVE_HEADER)
        message(FATAL_ERROR "Missing system header ${HEADER}")
    endif()
endfunction()

# sendmsg, recvmsg, bunch of other stuff
require_system_header("sys/socket.h")
```

That's a little bigger change than I anticipated; if you don't want to do it in this PR I could do it afterwards.

**Comment by JosiahWI** [comment]
Sorry, I didn't read carefully, that only works for sendmsg and recvmsg. I think for recvmmsg and sendmmsg we need to check for the function, which I have not figured out how to do (autotools has its pros...). I would ignore this for now.

File: iocore/net/UnixUDPNet.cc [hunk 1]
```diff
@@ -326,61 +415,54 @@ UDPNetProcessorInternal::udp_read_from_net(UDPNetHandler *nh, UDPConnection *xuc
...
 #endif
-#if defined(IPV6_PKTINFO) || defined(IPV6_RECVPKTINFO)
-      case IPV6_PKTINFO: // IPV6_RECVPKTINFO uses IPV6_PKTINFO too
-        if (cmsg->cmsg_level == IPPROTO_IPV6) {
-          struct in6_pktinfo *pktinfo = reinterpret_cast<struct in6_pktinfo *>(CMSG_DATA(cmsg));
-          memcpy(toaddr.sin6_addr.s6_addr, &pktinfo->ipi6_addr, 16);
+    }
+
+    // If we got the gso size, then we need to find out in how many parts this was spliced.
+    int const parts = gso_size ? static_cast<int>(ceil(static_cast<double>(r) / static_cast<double>(gso_size))) : 1;
```
Comment (brbzull0) [inline iocore/net/UnixUDPNet.cc:434]: Not sure how bad would be using `ceil` just right here.

File: iocore/net/UnixUDPNet.cc [hunk 2]
```diff
@@ -1452,13 +1709,22 @@ net_signal_hook_callback(EThread *thread)
...
   SET_HANDLER(&UDPNetHandler::startNetEvent);
 }
 
+bool
+UDPNetHandler::is_gro_enabled() const
+{
+#ifndef SOL_UDP
+  return false;
+#endif
+  return this->_cfg.enable_gro;
```
Comment (maskit) [inline iocore/net/UnixUDPNet.cc:1725]: It may be ok, but some analyzer may say this line is unreachable. I'd have this line in an else clause.

Comment (brbzull0) [reply iocore/net/UnixUDPNet.cc:1725]: This is ugly, I meant to have an `else` here. Thanks

File: iocore/net/UnixUDPNet.cc (hunk 1 above)
Comment (brbzull0) [reply iocore/net/UnixUDPNet.cc:434]: We can have a remainder bigger than 0.

Comment (brbzull0) [reply iocore/net/UnixUDPNet.cc:434]: > Oh, you mean `r` can be 2500 (1000 + 1000 + 500) where `gso_size` is 1000?
> 

Exactly.


> I don't have any better idea to calculate parts faster.
> https://quick-bench.com/q/IC9D3TNEcLCwWgEfoh_Ll3igs2k

does not seems to  change much I think.


> Another approach would be get rid of parts. Maybe we can subtract gso_size from r on every iteration instead (gso_size > would need to be set to r if gso_size is 0)?

I could probably do that I think.

**Comment by brbzull0** [comment]
[approve ci autest]

**Comment by brbzull0** [comment]
[approve ci autest]

**Comment by brbzull0** [comment]
removed the `ceil` function and the parts handling, it now relies on the received size and work out with the remining size based on each part/packet size (if gro).

cc @maskit

File: iocore/net/UnixUDPNet.cc [hunk 3]
```diff
@@ -37,20 +37,32 @@
 #include "P_DNSConnection.h"
 #include "P_Net.h"
 #include "P_UDPNet.h"
+#include "tscore/ink_inet.h"
 
 #include "tscore/ink_sock.h"
-
-#include "netinet/udp.h"
+#include <math.h>
```
[the remark this answers is not in the document]
Comment (brbzull0) [reply iocore/net/UnixUDPNet.cc:43]: defintelly not. Thanks.

**Review by maskit — APPROVED** [review]
Looks good. Thanks.

**Pull request merged**